### PR TITLE
fix problem waiting for candidates

### PR DESCRIPTION
**Pull request by kalantar**
Observed that if the candidate was missing the experiment did not start when it became available.

### CODE REVIEW

**Comment by kalantar** [comment]
@kimikowang I would like your feedback on this solution. I found that `detectTargets()` was returning err when the targets were missing. In response, `syncKubernetes()` would end the experiment (via `endRequest()`). I modified the return values so that it could requue the request instead.

I then found that validateDetectedRules() would fail because it expected label `iter8-tools/role` to be `progressing` or `stable`. However, when the targets have not yet been identified this label had not been set on the dr. I introduced a new value `initializing` that could be set earlier and modified the validation.

**Comment by kalantar** [comment]
I've marked this as work in progress while we discuss this approach. If you think the approach is ok, I will mark it ready for review and we can proceed.

**Comment by kimikowang** [comment]
> @kimikowang I would like your feedback on this solution. I found that `detectTargets()` was returning err when the targets were missing. In response, `syncKubernetes()` would end the experiment (via `endRequest()`). I modified the return values so that it could requue the request instead.
> 
> I then found that validateDetectedRules() would fail because it expected label `iter8-tools/role` to be `progressing` or `stable`. However, when the targets have not yet been identified this label had not been set on the dr. I introduced a new value `initializing` that could be set earlier and modified the validation.

What is the purpose of requeue the request if targets are missing?

**Comment by kalantar** [comment]
> What is the purpose of requeue the request if targets are missing?

Maybe my misunderstanding about how the controller is triggered. I see that it works with just and `endRequest` Looks like the status message is wrong however. Let me fix.

**Comment by kalantar** [comment]
@kimikowang I've addressed your concern above. Thanks.

File: pkg/controller/experiment/routing/router.go [hunk 1]
```diff
@@ -172,7 +172,7 @@ func candiateSubsetName(idx int) string {
 	return SubsetCandidate + "-" + strconv.Itoa(idx)
 }
 
-func (r *Router) UpdateCandidates(targets *targets.Targets) (err error) {
+func (r *Router) UpdateCandidates(instance *iter8v1alpha2.Experiment, targets *targets.Targets) (err error) {
```
Comment (kimikowang) [inline pkg/controller/experiment/routing/router.go:175]: Is it necessary to introduce this input parameter?

File: pkg/controller/experiment/reconcile.go [hunk 2]
```diff
@@ -70,56 +70,56 @@ func (r *ReconcileExperiment) checkOrInitRules(context context.Context, instance
...
 			r.markRoutingRulesError(context, instance, "Fail in updating routing rule: %v", err)
-			return err
+			return false, err
 		}
 	}
 
 	r.markTargetsFound(context, instance, "")
 
-	return nil
+	return true, nil
```
Comment (kimikowang) [inline pkg/controller/experiment/reconcile.go:122]: After the fix, looks like the boolean/error pair is a bit redundant?

**Comment by kalantar** [comment]
@kimikowang I've added those changes to support this line: https://github.com/kalantar/iter8-controller/blob/missingCandidate/pkg/controller/experiment/kubernetes.go#L42 which requires both `err` and `found`. When there is an error, found may differ. I think the code changes are needed.

**Comment by fabolive** [comment]
@kimikowang: Is this ready to be merged?

**Comment by kalantar** [comment]
Oh I see the extra parameter is not what I thought; I will fix asap.

**Comment by kalantar** [comment]
removed the extra parameter

**Comment by fabolive** [comment]
@kimikowang: Your two remaining comments seem to have been addressed by the last 2 sets of changes.

**Comment by kimikowang** [comment]
> @kimikowang: Your two remaining comments seem to have been addressed by the last 2 sets of changes and explanation.

Let's merge it as you approved it. LGTM as well.

**Pull request merged**